### PR TITLE
Issue/day of month suffix

### PR DESCRIPTION
**Pull request by IgnacioCarrionN**
See #150

### CODE REVIEW

File: klock/src/commonMain/kotlin/com/soywiz/klock/PatternDateFormat.kt [hunk 1]
```diff
@@ -60,7 +60,7 @@ data class PatternDateFormat @JvmOverloads constructor(
             }
             val chunk = s.readChunk {
                 val c = s.readChar()
-                while (s.hasMore && s.tryRead(c)) Unit
+                while (s.hasMore && (s.tryRead(c) || s.tryReadDaySuffix(c))) Unit
```
Comment (soywiz) [inline klock/src/commonMain/kotlin/com/soywiz/klock/PatternDateFormat.kt:63]: Why is this change necessary? Is not covered multi-character patterns already somehow?

File: klock/src/commonMain/kotlin/com/soywiz/klock/PatternDateFormat.kt [hunk 2]
```diff
@@ -126,6 +127,11 @@ data class PatternDateFormat @JvmOverloads constructor(
 	// EEE, dd MMM yyyy HH:mm:ss z -- > Sun, 06 Nov 1994 08:49:37 GMT
     // YYYY-MM-dd HH:mm:ss
 
+    val suffixes = listOf("0th",  "1st",  "2nd",  "3rd",  "4th",  "5th",  "6th",  "7th",  "8th",  "9th",
+        "10th", "11th", "12th", "13th", "14th", "15th", "16th", "17th", "18th", "19th",
+        "20th", "21st", "22nd", "23rd", "24th", "25th", "26th", "27th", "28th", "29th",
+        "30th", "31st")
```
Comment (soywiz) [inline klock/src/commonMain/kotlin/com/soywiz/klock/PatternDateFormat.kt:133]: I think this should be part of the locale, since this is different on each language. I think as a first approach we can provide English only, then request people from different countries to help us adjusting this. Shouldn't be a big deal making this incrementally since people is not using this yet

Comment (soywiz) [reply klock/src/commonMain/kotlin/com/soywiz/klock/PatternDateFormat.kt:133]: Additionally I believe the name could be improved. Since the list is not a list of suffixes, but an actual list of ordinals (if I remember correctly).

Comment (soywiz) [reply klock/src/commonMain/kotlin/com/soywiz/klock/PatternDateFormat.kt:133]: When moving to the locale I see that this can be a table generated. Not sure if output side would be better with a generated table or like that, but maybe:

```
when (num % 10) {
    1 -> "st"
    2 -> "nd"
    3 -> "rd"
    else -> "th"
}
```

could be checked faster against the actual rules and should be less prone to errors, still I'm also ok with the precomputed table as long as it is part of the locale, and the precomputed tables works for all the languages (for example not depending on the month, etc.)

I'm not fully sure if the implementation is using an Appender right now, but another performance-wise option without a table could be similar this too:

```
// English
fun writeOrdinal(day: Int, out: Appender) {
       out.append(day)
       when (num) {
             ....
       }
}
```

I'll let you decide how to do this. And other people feedback in this area is also welcome since I have limited info here to provide better feedback!

**Review by soywiz — COMMENTED** [review]
Nice PR!
Can you help me to understand the read stuff, and move the ordinal stuff to a way where it can be internationalized?

File: klock/src/commonMain/kotlin/com/soywiz/klock/PatternDateFormat.kt (hunk 2 above)
Comment (IgnacioCarrionN) [reply klock/src/commonMain/kotlin/com/soywiz/klock/PatternDateFormat.kt:133]: I used a precomputed table because it's easier to translate to other languages. I'm not sure if any language use different suffixes for each day. If we use a precomputed table or not I think it should be a list with 31 elements to keep the implementation simple and less prone to errors.

File: klock/src/commonMain/kotlin/com/soywiz/klock/internal/MicroStrReader.kt [hunk 3]
```diff
@@ -21,6 +21,13 @@ internal class MicroStrReader(val str: String, var offset: Int = 0) {
 		readChar()
 		return true
 	}
```
Comment (IgnacioCarrionN) [inline klock/src/commonMain/kotlin/com/soywiz/klock/internal/MicroStrReader.kt:23]: In the implementation of tryRead if the next char it's not equal to the char provided to the method returns false and doesn't read the char. For this case I used another function to read if the current char is 'd' and the next char is 'o'.

File: klock/src/commonMain/kotlin/com/soywiz/klock/PatternDateFormat.kt (hunk 1 above)
Comment (IgnacioCarrionN) [reply klock/src/commonMain/kotlin/com/soywiz/klock/PatternDateFormat.kt:63]: `fun tryRead(expected: Char): Boolean {
		if (eof || peekChar() != expected) return false  
		readChar()  
		return true  
}`
In s.tryRead(current) if "expected" is not equal to the next character returns false and doesn't call to readChar(). 

`fun tryReadDaySuffix(current: Char): Boolean {
        if (current == 'd' && peekChar() == 'o') {
            readChar()
            return true
        }
        return false
}`

For this case I used tryReadDaySuffix(current) where if "current" is equal to "d" and next char is equal to "o" calls readChar()

File: klock/src/commonMain/kotlin/com/soywiz/klock/PatternDateFormat.kt (hunk 2 above)
Comment (soywiz) [reply klock/src/commonMain/kotlin/com/soywiz/klock/PatternDateFormat.kt:133]: What about (on the Locale class):

```
override protected val ordinals = Array(32) {
  when (it % 10) {
    1 -> "${it}st"
    2 -> "${it}nd"
    3 -> "${it}rd"
    else -> "${it}th"
  }
}
open fun getOrdinalByDay(day: Int): String = ordinals[day] // Just in case someone doesn't want to use a table
```

Then on a test (or similar):

```
assertEquals("0th 1st 2nd", locale.ordinals.joinToString(" "))
```

That way we are providing the rules on the implementation, yet we can check each value independently for errors or as a reference on the test. By being an `open val`, other languages can either provide a rule-based orginal generation, or a precomputed literal list with: `override val ordinals = arrayOf("", ...)` each language can choose what fits better for their rules.

Still I let you decide what to do here. I'm ok with literals or rules, as long as it supports localization

File: klock/src/commonMain/kotlin/com/soywiz/klock/PatternDateFormat.kt (hunk 1 above)
Comment (soywiz) [reply klock/src/commonMain/kotlin/com/soywiz/klock/PatternDateFormat.kt:63]: What about this?

```diff
Index: klock/src/commonMain/kotlin/com/soywiz/klock/PatternDateFormat.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- klock/src/commonMain/kotlin/com/soywiz/klock/PatternDateFormat.kt	(revision b730afdf20fa13b7792c4a078f5e3e81d78ce862)
+++ klock/src/commonMain/kotlin/com/soywiz/klock/PatternDateFormat.kt	(date 1604673400838)
@@ -58,9 +58,9 @@
                     continue
                 }
             }
-            val chunk = s.readChunk {
+            val chunk = s.tryReadOrNull("do") ?: s.readChunk {
                 val c = s.readChar()
-                while (s.hasMore && (s.tryRead(c) || s.tryReadDaySuffix(c))) Unit
+                while (s.hasMore && (s.tryRead(c))) Unit
             }
             chunks.add(chunk)
         }
Index: klock/src/commonMain/kotlin/com/soywiz/klock/internal/MicroStrReader.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- klock/src/commonMain/kotlin/com/soywiz/klock/internal/MicroStrReader.kt	(revision b730afdf20fa13b7792c4a078f5e3e81d78ce862)
+++ klock/src/commonMain/kotlin/com/soywiz/klock/internal/MicroStrReader.kt	(date 1604673389034)
@@ -21,19 +21,15 @@
 		readChar()
 		return true
 	}
-    fun tryReadDaySuffix(current: Char): Boolean {
-        if (current == 'd' && peekChar() == 'o') {
-            readChar()
-            return true
-        }
-        return false
-    }
     fun tryRead(expected: String): Boolean {
         if (expected.length > available) return false
         for (n in expected.indices) if (this.str[offset + n] != expected[n]) return false
         offset += expected.length
         return true
     }
+    fun tryReadOrNull(expected: String): String? {
+        return if (tryRead(expected)) expected else null
+    }
     fun read(count: Int): String = this.str.substring(offset, (offset + count).coerceAtMost(length)).also { this.offset += it.length }
 	fun readRemaining(): String = read(available)
     fun readInt(count: Int): Int = read(count).toInt()
```

I think that tryReadDaySuffix doesn't belong to MicroStrReader. That class is not coupled at all to klock, it is a generic StrReader.
That function could be provided as a private extension function, but I think you don't want to try to read the day suffix on each iteration.

Comment (soywiz) [reply klock/src/commonMain/kotlin/com/soywiz/klock/PatternDateFormat.kt:63]: In fact I think that would read better with this:

```
private fun MicroStrReader.readRepeatedChar(): String {
    return readChunk {
        val c = readChar()
        while (hasMore && (tryRead(c))) Unit
    }
}
```

```
chunks.add(s.tryReadOrNull("do") ?: s.readRepeatedChar())
```

Because I had to figure out what that loop did when I read it since the time it was written

Comment (IgnacioCarrionN) [reply klock/src/commonMain/kotlin/com/soywiz/klock/PatternDateFormat.kt]: Done

File: klock/src/commonMain/kotlin/com/soywiz/klock/PatternDateFormat.kt (hunk 2 above)
Comment (IgnacioCarrionN) [reply klock/src/commonMain/kotlin/com/soywiz/klock/PatternDateFormat.kt]: Done as open val ordinals and open functions getOrdinalByDay and getDayByOrdinal.

File: klock/src/commonTest/kotlin/com/soywiz/klock/locale/KlockLocaleTest.kt [hunk 4]
```diff
@@ -276,4 +276,17 @@ class KlockLocaleTest {
 			HttpDate.withTimezoneNames(ExtendedTimezoneNames).parse("Tue, 19 Sep 2017 00:58:45 MSK").toString()
 		)
 	}
+
+    @Test
+    fun testOrdinalsEnglish() {
```
Comment (soywiz) [inline klock/src/commonTest/kotlin/com/soywiz/klock/locale/KlockLocaleTest.kt:281]: I think better to test the 31 days at once: it is failing for 11, 12 and 13. In fact, the code I put was incomplete since those are irregulars.

File: klock/src/commonMain/kotlin/com/soywiz/klock/PatternDateFormat.kt (hunk 2 above)
Comment (soywiz) [reply klock/src/commonMain/kotlin/com/soywiz/klock/PatternDateFormat.kt]: The code I put is wrong, since there are three irregulars: 11, 12, 13. Can you fix it?

Comment (IgnacioCarrionN) [reply klock/src/commonMain/kotlin/com/soywiz/klock/PatternDateFormat.kt]: Done

File: klock/src/commonTest/kotlin/com/soywiz/klock/locale/KlockLocaleTest.kt (hunk 4 above)
Comment (IgnacioCarrionN) [reply klock/src/commonTest/kotlin/com/soywiz/klock/locale/KlockLocaleTest.kt:281]: Done

Comment (soywiz) [reply klock/src/commonTest/kotlin/com/soywiz/klock/locale/KlockLocaleTest.kt:281]: I usually prefer using string literals and joining to string so you can copy paste the results when changed.
Still this should do

**Review by soywiz — APPROVED** [review]
Thank you very much for your nice work here!

**Comment by soywiz** [comment]
Fixes #150
